### ISSUE TITLE
Add query clause for inner joins

### ISSUE THREAD

**Issue by gimantha** [ballerina-platform/ballerina-lang#22382]
A join-clause does an inner equijoin, and implements it as a hash join.

```
intermediate-clause :=
   ...
   | join-clause

join-clause :=
   "join" typed-binding-pattern "in" expression join-on-condition
join-on-condition := "on" expression "equals" expression
```

Consider

```
from X in E1
  join Y in E2 on E3 equals E4
  S
```

The scope of bindings is as follows:

*   bindings from X are in scope in E3 and S but not E4 or E2
*   bindings from Y are in scope in E4 and S but not E3

A join-clause is executed as follows:

1. Evaluate the expression following "in" (with nothing extra in scope) resulting in an iterable value c 
    1. create an empty hash table that allows multiple entries per key and uses == equality
    2. for each item in c
        1. bind c to typed-binding-pattern resulting in frame f'
        2. evaluate the RHS of "equals" with f' in scope resulting in value k'
        3. add an entry into the hash table from k' to f'
    3. for each input frame f
        4. evaluate the LHS of "equals" with f in scope resulting in value k
        5. for each frame f' with key k in the hash table
            1. emit f augmented with f'

**Comment by grainier** [comment]
Will get merged in #23237

**Comment by grainier** [comment]
https://github.com/grainier/ballerina/pull/33
